src/OrchardCore.Modules/OrchardCore.Setup/Filters/AutoSetupStartupFilter.cs

```diff
@@ -0,0 +1,164 @@
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Data;
+using System.Data.Common;
+using System.Linq;
+using System.Text;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Localization;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Environment.Shell.Scope;
+using OrchardCore.Setup.Options;
+using OrchardCore.Setup.Services;
+
+namespace OrchardCore.Setup
+{
+    public class AutoSetupStartupFilter : IStartupFilter
+    {
+        private readonly IServiceProvider _provider;
+        private readonly IShellHost _shellHost;
+        private readonly ILogger _logger;
+        private readonly IStringLocalizer T;
+
+        public AutoSetupStartupFilter(IServiceProvider provider, IShellHost shellHost, IStringLocalizer<AutoSetupStartupFilter> stringLocalizer, ILogger<AutoSetupStartupFilter> logger)
+        {
+            T = stringLocalizer;
+            _provider = provider;
+            _shellHost = shellHost;
+            _logger = logger;
+        }
+
+        public Action<IApplicationBuilder> Configure(Action<IApplicationBuilder> next)
+        {
+            var shellServices = ShellScope.Context.ServiceProvider;
+            var scopedServices = ShellScope.Services;
+            var settings = ShellScope.Context.Settings;
+            var state = ShellScope.Context.Settings.State;
+
+            var siteIsUninitialized = state == OrchardCore.Environment.Shell.Models.TenantState.Uninitialized;
+
+            if (siteIsUninitialized)
+            {
+                var optionsAccessor = shellServices.GetRequiredService<IOptions<AutoSetupOptions>>();
+                var options = optionsAccessor.Value;
+
+                if (options != null)
+                {
+                    _logger.LogInformation("AutoSetup is initializing the site");
+                    var validationContext = new ValidationContext(options, _provider, null);
```

Comment:
Here you can use `services` in place of `_provider` that you don't need to inject in the ctor